File: client/go/message.go
package conduitclient

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/google/uuid"
)

// MessageType represents the type of message
type MessageType string

const (
	// MessageTypeData is a regular data message
	MessageTypeData MessageType = "data"

	// MessageTypeControl is a control plane message
	MessageTypeControl MessageType = "control"

	// MessageTypeError is an error message
	MessageTypeError MessageType = "error"
)

// Message represents a message in the Exchange
type Message struct {
	// ID is a unique identifier for this message
	ID string `json:"id"`

	// Timestamp is when the message was created
	Timestamp time.Time `json:"timestamp"`

	// Sequence is the NATS stream sequence number
	Sequence uint64 `json:"sequence"`

	// Type is the message type
	Type MessageType `json:"type"`

	// Payload is the message content (arbitrary JSON)
	Payload json.RawMessage `json:"payload"`

	// Metadata contains optional metadata
	Metadata map[string]string `json:"metadata,omitempty"`
}

// UnmarshalPayload unmarshals the payload into the given struct
func (m *Message) UnmarshalPayload(v interface{}) error {
	return json.Unmarshal(m.Payload, v)
}

// NewMessage creates a new message with the given payload
func NewMessage(msgType MessageType, payload interface{}) (*Message, error) {
	payloadBytes, err := json.Marshal(payload)
	if err != nil {
		return nil, err
	}

	return &Message{
		ID:        uuid.New().String(),
		Timestamp: time.Now(),
		Type:      msgType,
		Payload:   payloadBytes,
		Metadata:  make(map[string]string),
	}, nil
}

// NewDataMessage creates a new data message (convenience function)
func NewDataMessage(payload interface{}) (*Message, error) {
	return NewMessage(MessageTypeData, payload)
}

// MarshalMessage marshals a message to JSON bytes
func MarshalMessage(msg *Message) ([]byte, error) {
	return json.Marshal(msg)
}

// UnmarshalMessage unmarshals a message from JSON bytes
func UnmarshalMessage(data []byte) (*Message, error) {
	var msg Message
	if err := json.Unmarshal(data, &msg); err != nil {
		return nil, err
	}

	// Validate the message
	if err := ValidateMessage(&msg); err != nil {
		return nil, err
	}

	return &msg, nil
}

// ValidateMessage validates that a message has all required fields
func ValidateMessage(msg *Message) error {
	if msg.ID == "" {
		return fmt.Errorf("message ID is required")
	}
	if msg.Type == "" {
		return fmt.Errorf("message type is required")
	}
	if msg.Type != MessageTypeData && msg.Type != MessageTypeControl && msg.Type != MessageTypeError {
		return fmt.Errorf("invalid message type: %s", msg.Type)
	}
	if len(msg.Payload) == 0 {
		return fmt.Errorf("message payload is required")
	}
	return nil
}

File: client/typescript/src/message.ts
/**
 * Message envelope format for Conduit.
 */

import { v4 as uuidv4 } from 'uuid';

/**
 * Message type enumeration.
 */
export enum MessageType {
  DATA = 'data',
  CONTROL = 'control',
  ERROR = 'error',
}

/**
 * Message envelope for Conduit exchanges.
 */
export interface Message<T = any> {
  /** Unique message identifier */
  id: string;

  /** ISO8601 timestamp when message was created */
  timestamp: string;

  /** NATS stream sequence number */
  sequence: number;

  /** Message type */
  type: MessageType;

  /** Message content (arbitrary JSON-serializable data) */
  payload: T;

  /** Optional metadata */
  metadata?: Record<string, string>;
}

/**
 * Create a new message with the given type and payload.
 */
export function createMessage<T>(
  type: MessageType,
  payload: T,
  metadata?: Record<string, string>
): Message<T> {
  return {
    id: uuidv4(),
    timestamp: new Date().toISOString(),
    sequence: 0,
    type,
    payload,
    metadata: metadata || {},
  };
}

/**
 * Create a new data message (convenience function).
 */
export function createDataMessage<T>(
  payload: T,
  metadata?: Record<string, string>
): Message<T> {
  return createMessage(MessageType.DATA, payload, metadata);
}

/**
 * Serialize a message to JSON bytes.
 */
export function serializeMessage<T>(msg: Message<T>): Uint8Array {
  const json = JSON.stringify(msg);
  return new TextEncoder().encode(json);
}

/**
 * Deserialize a message from JSON bytes.
 */
export function parseMessage<T>(data: Uint8Array): Message<T> {
  const json = new TextDecoder().decode(data);
  const msg = JSON.parse(json) as Message<T>;
  validateMessage(msg);
  return msg;
}

/**
 * Validate that a message has all required fields.
 *
 * @throws Error if message is invalid
 */
export function validateMessage<T>(msg: Message<T>): void {
  if (!msg.id) {
    throw new Error('message ID is required');
  }
  if (!msg.type) {
    throw new Error('message type is required');
  }
  if (
    msg.type !== MessageType.DATA &&
    msg.type !== MessageType.CONTROL &&
    msg.type !== MessageType.ERROR
  ) {
    throw new Error(`invalid message type: ${msg.type}`);
  }
  if (msg.payload === undefined || msg.payload === null) {
    throw new Error('message payload is required');
  }
}
